### PR TITLE
Try to recover from panic

### PR DESCRIPTION
**Pull request by mrsmkl**

### CODE REVIEW

**Comment by kobigurk** [comment]
The concept looks great!
Does `try_join_all` catch panics or only errors?

**Comment by mrsmkl** [comment]
> The concept looks great!
> Does `try_join_all` catch panics or only errors?

It catches panics from threads. You can check the previous commit for testing...

File: src/bin/contribute.rs [hunk 1]
```diff
@@ -1011,19 +1052,24 @@ impl Contribute {
                     )
                 }
             };
-
+            warn!("Moving chunk {} to upload", chunk_id);
```
Comment (kobigurk) [inline src/bin/contribute.rs]: Is this correct to be a `warn`? I'd think this is `debug` or `info`
Same about others

Comment (mrsmkl) [reply src/bin/contribute.rs]: I meant just to delete these, but looks like I missed them.

File: src/bin/contribute.rs [hunk 2]
```diff
@@ -382,66 +386,89 @@ impl Contribute {
...
+            match futures::future::try_join_all(futures).await {
+                Err(err) => {
+                    warn!("Crash!!!");
+                    // Afetr crashes have been incremented, one of the threads might have the PIPELINE lock.
+                    // But once that thread releases the lock, we clear the pipeline.
+                    // After that no old thread can add to pipeline lanes.
+                    // If the threads would test before they have the lock, the pipeline might be cleared just before they get the lock.
+                    // Then they could modify the pipeline and cause deadlock.
+                    CRASHES.fetch_add(1, SeqCst);
+                    self.clear_pipeline()?;
```
Comment (kobigurk) [inline src/bin/contribute.rs:456]: Can we reach a situation here where the following happens?

* A thread gets the write lock and goes over the CRASHES check
* CRASHES gets incremented
* We clear the pipelines
* The thread works on invalid data

If yes, maybe it'd be safer to increment CRASHES while we have the write lock?

Comment (mrsmkl) [reply src/bin/contribute.rs:456]: I think it's possible anyway that a thread will work on invalid data, I'll add something to make sure that the filenames won't conflict.

Comment (kobigurk) [reply src/bin/contribute.rs:456]: When I wrote "invalid data", I actually meant invalid pipeline data :)

**Comment by kobigurk** [comment]
@mrsmkl  The only thing that holds me up from merging is the complexity around pipeline clearing. I don't feel confident that there are no data races. Do you see any way we can simplify it?

EDIT: no longer relevant.

File: src/bin/contribute.rs (hunk 2 above)
Comment (mrsmkl) [reply src/bin/contribute.rs:456]: Well in any case the property we want to have is that after the pipelines have been cleared, the old threads cannot modify the PIPELINES data structure anymore. I think this holds if we increment the CRASHES counter before or during clearing the pipelines. Only difference should be that in theory more threads can modify the pipeline if we are waiting for the write lock before incrementing the counter.

Here's what should happen in your scenario
* A thread gets the write lock and goes over the CRASHES check
* CRASHES gets incremented
* We clear the pipelines -- clearing the pipelines requires write lock, so the thread has released it before this step is finished
* The thread works on invalid data -- the thread will exit when it tries to modify PIPELINE

Comment (kobigurk) [reply src/bin/contribute.rs:456]: Got it! That makes sense. Maybe the last thing we should add then is cleanup of older "crash level" files.

File: src/bin/contribute.rs [hunk 3]
```diff
@@ -382,65 +403,129 @@ impl Contribute {
...
-                                    .remove_chunk_id_from_lane_if_exists(
-                                        &PipelineLane::Upload,
-                                        &chunk_id,
-                                    )
-                                    .expect("Should have removed chunk ID from lane")
-                                {
-                                    let _ =
-                                        cloned.unlock_chunk(&chunk_id, Some(e.to_string())).await;
+
+        loop {
```
Comment (kobigurk) [inline src/bin/contribute.rs:383]: I've been testing this on a VM that's limited to 192kbps, and so times out often.
I reached a situation where I only have 1 chunk in the Download pipeline and it's stuck there.
Re-reading the code, I believe I understand what happens - only 2/3 async tasks exited, and so the outer loop is not invoked again.

Comment (mrsmkl) [reply src/bin/contribute.rs:383]: If it panics, it should always go to the outer loop. Other errors are handled in the inner loop. Some times will get stuck if the locks are not released on the server (the check for this is ran every 5 minutes), perhaps that's what happened?

Comment (kobigurk) [reply src/bin/contribute.rs:383]: You're right! I'll give a try to the updated code (if you can commit a version without the crashes it'd be nice)

**Comment by kobigurk** [comment]
Seems like we got by without it. Closing for now.

**Pull request closed without merging**